thorlcr/thorutil/thmem.cpp

```diff
@@ -229,6 +243,8 @@ class CSpillableStreamBase : public CSimpleInterface, implements roxiemem::IBuff
     }
     virtual bool freeBufferedRows(bool critical)
     {
+        if (spillFile) // if spilt already. NB: thread-safe, as only set by this thread and can't this freeBufferedRows() can't be called on multiple threads concurrently.
+            return false;
```

Comment:
@richardkchapman - NB: this is the actual deadlock fix, the other changes ensure callback removed in other cases and safeguard attempt to remove >1.